### ISSUE TITLE
style_file crashes because of docopt when processing single files with spaces in the path

### ISSUE THREAD

**Issue by kcphila** [lorenzwalthert/precommit#492]
https://github.com/lorenzwalthert/precommit/blob/a292fa679ab814eecf3a2c1905e4f04f872c48d4/inst/hooks/exported/style-files.R#L41

We've run into a problem with the `style-files` precommit hook on version 0.3.2 .

When a single file is passed into `style_file` that contains multiple spaces, the `arguments <- docopt::docopt(doc, args)` command separates the files at each space, leading to incorrect evaluations.

For example - the first 3 do just fine, then the bottom 2 show how this fails.

```R
 docopt::docopt(doc, 
     c("requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R", 
        "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact 2.R")
)$files
[1] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"  
[2] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact 2.R"

docopt::docopt(doc, c(
       "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R", 
       "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R")
)$files
[1] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"
[2] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"

docopt::docopt(doc, c(
     "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R", 
     "anything_else.R")
)$files
[1] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"
[2] "anything_else.R"                                                  

docopt::docopt(doc, c("requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"))$files
[1] "requests/OneOffDataRequests/drug_delivery/func/pull"
[2] "arrest"                                             
[3] "fact.R"                                             

docopt::docopt(doc, c("pull arrest fact.R"))$files
[1] "pull"   "arrest" "fact.R"
``` 

When attempting to commit files in a real repo, the precommit hook fails as so:

```
[1] "cache root set to " "styler-perm"
Styling  4  files:
 requests/One                                  Error in `transform_code()`:
! requests/One is not an R, Rmd or Rnw file
Backtrace:
     ▆
  1. ├─base::tryCatch(...)
  2. │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
  3. │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
  4. │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
  5. ├─base::do.call(style_file, dot_args)
  6. └─styler (local) `<fn>`(path = `<chr>`, style = `<fn>`)
  7.   └─styler:::transform_files(...)
  8.     └─purrr::map_lgl(...)
  9.       └─styler (local) .f(.x[[i]], ...)
 10.         └─styler:::transform_code(path, fun = fun, ..., dry = dry)
 11.           └─rlang::abort(paste(path, "is not an R, Rmd or Rnw file"))
Execution halted

```

I'm not familiar with docopt, but it seems `style_files.R` relies on it pretty heavily and there may be cases when this is correct for docopt but I am not sure if there are cases when it will be correct for `style_files.R` 

It looks like, tentatively, that appending the empty string will make this a non issue:

```R
docopt::docopt(doc, c("pull arrest fact.R", ""))$files
[1] "pull arrest fact.R"

Browse[2]> docopt::docopt(doc, c("requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R", ""))$files
[1] "requests/OneOffDataRequests/drug_delivery/func/pull arrest fact.R"```
```

Therefore, I think this code will fix the precommit hook - at least it passes with the files we're dealing with. I don't know if there are other side effects to be aware of.

```R

if(length(args) == 1){
  arguments <- docopt::docopt(doc, c(args, ""))
}else{
  arguments <- docopt::docopt(doc, args)
}
```

**Comment by lorenzwalthert** [comment]
Thanks for the detailed report. But file names with spaces is widely considered an anti-pattern. It brings all kinds of problems. On command line, spaces are used to separate arguments unless you quote them. Like `executable-name „file 1“ „file 2“`. Can you check docopt grasps that? Probably you will need an end to end example. If this works, then the pre-commit upstream framework might be the problem. In any case, I consider this out of scope for {precommit} and I strongly advice you to remove spaces and other non alpha numeric characters from your file names.

**Comment by kcphila** [comment]
Hi @lorenzwalthert ,

I'm not in control of the file naming, just supporting the analysts.  For all major languages, including bash, python, and R, the quotes of a script parameter on the commandline are stripped before they are passed into the arg variable of the script.

But to demonstrate, here's a trivial example demonstrate the problem:

```bash
mkdir break-precommit
cd break-precommit/

echo -e "repos:\n- repo: https://github.com/lorenzwalthert/precommit\n  rev: v0.3.2\n\
  hooks:\n      - id: style-files" >>  .pre-commit-config.yaml

touch "this filename is ill-advised.R"
git init 
git add .pre-commit-config.yaml
git add "this filename is ill-advised.R"
git commit -m "broken"
```

**Comment by lorenzwalthert** [comment]
Ok, so basically you are saying that {docopt}'s parsing fails for one argument with spaces in it 

Create `test-script.R`
```r
#!/usr/bin/env Rscript
'style files.
Usage:
  style_files <files>...

' -> doc

library(styler)
args <- commandArgs(trailingOnly = TRUE)

docopt::docopt(doc, args)
```

```bash 
test-script.R "a is here" "b"
List of 2
 $ <files>: chr [1:2] "a is here" "b"
 $ files  : chr [1:2] "a is here" "b"
NULL
```
```bash
test-script.R "a is here"    
List of 2
 $ <files>: chr [1:3] "a" "is" "here"
 $ files  : chr [1:3] "a" "is" "here"
NULL
```

This is IMO a bug in {docopt} and should be fixed there.

**Comment by kcphila** [comment]
> This is IMO a bug in {docopt} and should be fixed there. I recommend to open an issue there.

I don't think so - looking at the [docopt.R code repo](https://github.com/docopt/docopt.R), it looks like it is specifically designed to either accept a vector of command line arguments *or* a single line that includes all arguments, and since R doesn't distingush between `"string or length-1vector"` and `c("string or length-1vector")`, it's assuming you want the first.  

Since it looks like you're just running `args <- commandArgs(trailingOnly = TRUE)` and that's effectively what `docopt` does when the `args` param is missing, I think you can just exclude the pre-fetching of the commandline and everything works.

I pulled the heading of `style_files.R` and changed the call to docopt to simply `arguments <- docopt::docopt(doc)`, it appears to process correctly.

```bash
Rscript test-docopt.R "file and friends.R" "woof and whatever"
  files: file and friends.R  
  files: woof and whatever

Rscript test-docopt.R "file and friends.R"
 files: file and friends.R

Rscript test-docopt.R "file and friends.R" --style_fun whatever
  --style_fun: whatever
  files: file and friends.R
```

**Comment by lorenzwalthert** [comment]
> I think you can just exclude the pre-fetching of the commandline and everything works.

I am not sure I get what you mean. 

It's also super weird that setting an argument explicitly to its default changes the behaviour as in `docopt::docopt(doc, args = commandArgs(TRUE))`. But I now see that probably the solution is what you suggested. Actually, what's the drawback of always adding `"`, i.e. 
```
docopt::docopt(doc, c(args, ""))
```
instead of 
```
if(length(args) == 1){
  arguments <- docopt::docopt(doc, c(args, ""))
} else {
  arguments <- docopt::docopt(doc, args)
}```

**Comment by kcphila** [comment]
> what's the drawback of always adding `"`, i.e.
> ```
> docopt::docopt(doc, c(args, ""))
> ```

That's more concise - and I don't think there would be a drawback there. 

I also agree it's weird that providing the argument explicitly alters behavior, but it seems consistent with the case when explicitly passing NULL in to a function with a default value sets the param to NULL and doesn't trigger the default. `docopt` does special processing if the `args` parameter is missing from the function call.

**Comment by lorenzwalthert** [comment]
Seems yes. And since we alter `args` in style-files, not providing a value is not an option.

**Comment by kcphila** [comment]
> @kcphila would you like to contribute the solution in a PR? And as a follow up, review all other hooks too and adapt them to support file names with spaces?

@lorenzwalthert - sure, I'd be happy to.

**Comment by lorenzwalthert** [comment]
@kcphila are you still interested in contributing?

**Comment by kcphila** [comment]
> in the next week or so.

Actually - I just finished it up and sent the PR.

**Comment by lorenzwalthert** [comment]
Closed with #516.
